internal/controller/config/config.go

```diff
@@ -0,0 +1,90 @@
+package config
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/spf13/viper"
+	corev1 "k8s.io/api/core/v1"
+)
+
+type ConfigManager struct {
+	config *GlobalConfig
+	lock   *sync.RWMutex
+}
+
+func NewConfigManager() *ConfigManager {
+	return &ConfigManager{
+		config: &GlobalConfig{},
+		lock:   new(sync.RWMutex),
+	}
+}
+
+// GlobalConfig is the configuration for the controllers, it is
+// supposed to be populated from the configmap attached to the
+// controller manager.
+type GlobalConfig struct {
+	ClusterName     string                    `json:"clusterName"`
+	TimeIntervalSec uint                      `json:"timeIntervalSec"`
+	RepoCredentials map[string]*GitCredential `json:"repoCredentials"`
+}
+
+type GitCredential struct {
+	HTTPCredential *HTTPCredential `json:"HTTPCredential"`
+	SSHCredential  *SSHCredential  `json:"SSHCredential"`
+	TLS            *TLS            `json:"TLS"`
+}
+
+type HTTPCredential struct {
+	Username string            `json:"username"`
+	Password SecretKeySelector `json:"password"`
+}
+
+type SSHCredential struct {
+	SSHKey SecretKeySelector `json:"SSHKey" yaml:"SSHKey" `
+}
+
+type TLS struct {
+	InsecureSkipVerify bool              `json:"insecureSkipVerify"`
+	CACertSecret       SecretKeySelector `json:"CACertSecret"`
+	CertSecret         SecretKeySelector `json:"CertSecret"`
+	KeySecret          SecretKeySelector `json:"keySecret"`
+}
+
+type SecretKeySelector struct {
+	corev1.LocalObjectReference `mapstructure:",squash"` // for viper to correctly parse the config
+	Key                         string                   `json:"key" `
+	Optional                    *bool                    `json:"optional,omitempty" `
+}
+
+func (cm *ConfigManager) GetConfig() *GlobalConfig {
+	cm.lock.RLock()
+	defer cm.lock.RUnlock()
+	return cm.config
```

Comment:
What I was thinking was that this should return a `GlobalConfig` as opposed to the pointer. If it returns the pointer, then it's returning a reference to what's stored within the `ConfigManager`, which could change on it, when it's in the middle of reading it. 

The idea is that any user of this would get a copy of the `GlobalConfig`, guaranteed to be atomically correct, periodically, if it needs to. Then it's guaranteed to be regularly refreshing to the latest, but at the same time getting a copy that's atomically correct.

Reply:
Why was this resolved @shubhamdixit863 ?

Reply:
But we would be needing a Deep copy method then ,as internal values are also pointers ? @juliev0 ,may be like a nested deep copy  ?

Reply:
> But we would be needing a Deep copy method then ,as internal values are also pointers ? @juliev0 ,may be like a nested deep copy ?

great call out there, yeah

Reply:
Yeah ,ok I have  to restructure controller tests too ,then.